torch/_inductor/lowering.py

```diff
@@ -3932,7 +3983,20 @@ def adaptive_max_pool2d(x, output_size):
         )
     if h_in % h_out == 0 and w_in % w_out == 0:
         kernel_size = [h_in // h_out, w_in // w_out]
-        return max_pool2d_with_indices(x, kernel_size)
+        if should_fallback_max_pool2d_with_indices(kernel_size, dilation=[1, 1]):
+            return max_pool2d_with_indices(x, kernel_size)  # type: ignore[name-defined]   # noqa: F821
+        else:
+            v, offsets = _low_memory_max_pool2d_with_offsets(
+                x,
+                kernel_size,
+                stride=kernel_size,
+                padding=[0, 0],
+                dilation=[1, 1],
+                ceil_mode=False,
+            )
+            indices = _low_memory_max_pool2d_offsets_to_indices(
+                offsets, kernel_size[1], w_in, kernel_size, padding=[0, 0]
+            )
```

Comment:
This isn't returning in the non-fallback case? Also having the two kernels in the lowering doesn't actually allow for the offsets to be saved for backward.

I would suggest adding an inductor-specific decomposition for these special cases that calls `max_pool2d_with_indices`.

Reply:
Fixed the missing return.

> Also having the two kernels in the lowering doesn't actually allow for the offsets to be saved for backward.

Yes, that is true. There no improvement for `adaptive_max_pool2d` from this PR, but there is also no penalty. 

> I would suggest adding an inductor-specific decomposition for these special cases that calls `max_pool2d_with_indices`.

I did try to set up an inductor decomposition for this, but since the `kernel_size` is shape dependent on this path you still need to handle these special cases in the lowering.

Given the complexity this seemed to require I planned to address it in a follow up.

Reply:
When would we actually expect to not want to save indices ? I don't know if we need to fix this.

Reply:
I think @peterbell10 is referring to saving the int8 offsets and recomputing the int64 indices.  

When the split happens at this level (inside the lowering) the partitioner won't be able to see a call to `inductor_prims._low_memory_max_pool2d_offsets_to_indices` which it can recompute on the backward so the int64 indices are always saved for `adaptive_max_pool2d`. 

I agree that we don't need to fix this. It would be nice to at some point, but we dont need to now. I nductor has no problem fusing the "extra" `offsets_to_indices` kernel with the first one, so this effectively generates the same code as it did when there was a call to the old lowering for `max_pool2d_with_indices` here.

Reply:
Sure we can do this in a follow up PR. I don't follow this point though:

> since the `kernel_size` is shape dependent on this path you still need to handle these special cases in the lowering.

You should be able to compute `kernel_size` from `x.shape` and `out_shape` in the same way the lowering does.

Reply:
Yes, but here we `evaluate_static_shape` beforehand. When this logic is 
 moved out to an inductor decomposition x.shape` is symbolic and evaluating the conditional will throw.

Reply:
>  x.shape` is symbolic and evaluating the conditional will throw.

It should just add a guard, no?

Reply:
We add guards on height and width of max_pool anyway